keras_cv/layers/object_detection/roi_generator.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Mapping
+from typing import Optional
+from typing import Tuple
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+
+
+class ROIGenerator(tf.keras.layers.Layer):
+    """
+    Generates region of interests (ROI, or proposal) from scores.
+
+    Mainly used in Region CNN (RCNN) networks.
+
+    This works for a multi-level input, both boxes and scores are dictionary
+    inputs with the same set of keys.
+
+    Users can configure topk and threshold differently in train and inference.
+
+    Users can choose to combine all levels if nms across all levels are desired.
+
+    The following steps are applied to pair of (boxes, scores):
+    1) pre_nms_topk scores and boxes sorted and selected per level
+    2) nms applied and selected post_nms_topk scores and rois per level
+    3) combined scores and rois across all levels
+    4) post_nms_topk scores and rois sorted and selected
+
+    Args:
+        bounding_box_format: a case-insensitive string.
+            For detailed information on the supported format, see the
+            [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+        pre_nms_topk_train: int. number of top k scoring proposals to keep before applying NMS in training mode.
+            When RPN is run on multiple feature maps / levels (as in FPN) this number is per
+            feature map / level.
+        nms_score_threshold_train: float. score threshold to use for NMS in training mode.
```

Comment:
Param order wise, should we pair all the train/test param together? eg pre_nms_topk_train followed by pre_nms_topk_test

Reply:
Two options that I see:
1. group all train args, then all test args, as users mostly care about one thing or another at one time
2. let all args be list of [train_arg, test_arg]

Personally I think 1 is better for KerasTuner.

Reply:
2 will be more confusing, let's stay with 1 for now.